dolphin/db/sqlalchemy/api.py

```diff
@@ -136,23 +150,37 @@ def storage_get(context, storage_id):
 
 def storage_get_all(context, marker=None, limit=None, sort_keys=None,
                     sort_dirs=None, filters=None, offset=None):
-    """Retrieves all storage devices."""
-
     this_session = get_session()
     this_session.begin()
-    # TODO: need to handle all input parameters
-    all_storages = this_session.query(Storage).all()
-    return all_storages
-
-
-def volume_create(context, values):
-    """Create a volume from the values dictionary."""
-    return NotImplemented
-
+    if not sort_keys:
+        sort_keys= ['created_at']
+    if not sort_dirs:
+        sort_dirs = ['desc']
+    this_session = get_session()
+    this_session.begin()
+    query = this_session.query(models.Storage)
+
+    if filters:
+
+        for attr, value in filters.items():
+            query = query.filter(getattr(models.Storage, attr).like("%%%s%%" % value))
+    try:
+        for (sort_key,sort_dir) in zip(sort_keys,sort_dirs ):
+            query = apply_sorting(models.Storage, query, sort_key, sort_dir)
+    except AttributeError:
```

Comment:
apply_sorting raises InvalidInput exception, you are catching AttributeError?

Reply:
yes, catching the AtrributeError and rasing it as invlidInput . InvalidInput will be captured by the caller,

Reply:
Sorry, didn't get it...we are catching the exception from apply_sorting, which throws "raise exception.InvalidInput(reason=msg)"  line#49
But you are catching AttributeError...

Reply:
As discussed, Even If catch NotFound exception here I have nothing to resolve here, so leaving it to the caller to handle it.. You can see v1/storages.py is catching this exception and raising a http exception.